Add module for private endpoint DNS zone groups

l3ender opened the pull request:
##### SUMMARY
This PR adds a new module for configuring private endpoint DNS zone groups and is comparable to the [similar function in azure-cli](https://docs.microsoft.com/en-us/cli/azure/network/private-endpoint/dns-zone-group?view=azure-cli-latest). This is necessary when adding a private endpoint to a resource so that the private DNS entry will be used for the resource from within the vnet.

Consider the following example, where a Postgres server has already been created and added to an existing virtual network. The endpoint can be created:
```yaml
- name: Create private endpoint
  azure_rm_privateendpoint:
    name: my-private-endpoint
    resource_group: my-resource-group
    private_link_service_connections:
      - name: my-postgres-link
        private_link_service_id: "{{ postgres_server_output.id }}"
        group_ids:
          - postgresqlServer
```

However, the above by itself does not assign a private IP address to the Postgres server. The following must be done in order for that to occur:

```yaml
- name: Create private DNS zone for postgres
  azure_rm_privatednszone:
    name: privatelink.postgres.database.azure.com
    resource_group: my-resource-group

- name: Create virtual network link
  azure_rm_privatednszonelink:
    name: my-private-link
    resource_group: my-resource-group
    zone_name: privatelink.postgres.database.azure.com
    virtual_network: my-vnet-name

- name: Create zone group for private endpoint
  azure_rm_privateendpointdnszonegroup:
    name: my-zone-group
    private_endpoint: my-private-endpoint
    resource_group: my-resource-group
    private_dns_zone_configs:
      - name: default
        private_dns_zone: privatelink.postgres.database.azure.com
```

Here is output from a sample usage of the `azure_rm_privateendpointdnszonegroup` module:
```yaml
{
    "changed": true,
    "invocation": {
        "module_args": {
            "ad_user": null,
            "adfs_authority_url": null,
            "api_profile": "latest",
            "auth_source": "auto",
            "cert_validation_mode": null,
            "client_id": null,
            "cloud_environment": "AzureCloud",
            "log_mode": null,
            "log_path": null,
            "name": "zone-group-ef45326965",
            "password": null,
            "private_dns_zone_configs": [
                {
                    "name": "default",
                    "private_dns_zone_id": "/subscriptions/xxx/resourceGroups/automated-testing/providers/Microsoft.Network/privateDnsZones/privatelink.postgres.database.azure.com"
                }
            ],
            "private_endpoint": "privateendpoint-ef45326965",
            "profile": null,
            "resource_group": "automated-testing",
            "secret": null,
            "state": "present",
            "subscription_id": null,
            "tenant": null
        }
    },
    "state": {
        "id": "/subscriptions/xxx/resourceGroups/automated-testing/providers/Microsoft.Network/privateEndpoints/privateendpoint-ef45326965/privateDnsZoneGroups/zone-group-ef45326965",
        "name": "zone-group-ef45326965",
        "private_dns_zone_configs": [
            {
                "name": "default",
                "private_dns_zone_id": "/subscriptions/xxx/resourceGroups/automated-testing/providers/Microsoft.Network/privateDnsZones/privatelink.postgres.database.azure.com",
                "record_sets": [
                    {
                        "fqdn": "postgresqlsrv-ef45326965.privatelink.postgres.database.azure.com",
                        "ip_addresses": [
                            "10.1.0.4"
                        ],
                        "provisioning_state": "Succeeded",
                        "record_set_name": "postgresqlsrv-ef45326965",
                        "record_type": "A",
                        "ttl": 10
                    }
                ]
            }
        ],
        "provisioning_state": "Succeeded"
    }
}
```

Within the vnet, the hostname `postgresqlsrv-ef45326965.postgres.database.azure.com` can now be queried and it will return the private IP address for the Postgres server:
```bash
-> nslookup postgresqlsrv-ef45326965.postgres.database.azure.com
Server:		127.0.0.53
Address:	127.0.0.53#53

Non-authoritative answer:
postgresqlsrv-ef45326965.postgres.database.azure.com	canonical name = postgresqlsrv-ef45326965.privatelink.postgres.database.azure.com.
Name:	postgresqlsrv-ef45326965.privatelink.postgres.database.azure.com
Address: 10.1.0.4
```

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
azure_rm_privateendpointdnszonegroup
azure_rm_privateendpointdnszonegroup_info

##### ADDITIONAL INFORMATION

I found similar type of configuration in the `azure_rm_privateendpoint` module, but it was not functional and the private endpoint SDK did not support it. I have removed it to clarify usage.

Test coverage has been added (and acts as an example) in `tests/integration/targets/azure_rm_privateendpointdnszonegroup/tasks/main.yml`. It can be tested with the following playbook:

```yaml
---
- name: "Playbook for testing."
  hosts: "localhost"
  connection: "local"
  gather_facts: false
  vars:
    resource_group: "automated-testing"
    resource_group_secondary: "automated-testing-secondary"
  collections:
    - azure.azcollection

  tasks:
    - name: "Include tests"
      include_tasks: "tests/integration/targets/azure_rm_privateendpointdnszonegroup/tasks/main.yml"
```

l3ender committed:
add new modules for private endpoint dns zone groups

l3ender committed:
add test coverage for new modules

l3ender committed:
remove incorrect and unused config from module

l3ender committed:
add vnet link to private zone for full endpoint config

l3ender committed:
reorder test suite for clarity

l3ender commented:
Hello @Fred-sun, here is the PR I mentioned for private endpoint. It turns out no update to `azure.mgmt.network` dependency was required.

Please review and advise--thank you!

Fred-sun commented:
@l3ender Okay, we'll move forward with the merger as soon as possible. Thank you for your contribution!

l3ender committed:
Merge branch 'dev' into private-endpoint-dns-zones

Fred-sun commented on plugins/modules/azure_rm_privateendpointdnszonegroup.py:
```diff
@@ -0,0 +1,351 @@
+        self.to_do = Actions.NoAction
+
+        super(AzureRMPrivateEndpointDnsZoneGroup, self).__init__(self.module_arg_spec,
+                                                                 supports_tags=False,
```
Why do you inherit tags in your document, but set false here in argc? It is recommended that you add it here. And make the following changes. Thank you very much!
```suggestion
                                                                 supports_tags=True,
```

Fred-sun commented on plugins/modules/azure_rm_privateendpointdnszonegroup.py:
```diff
@@ -0,0 +1,351 @@
+
+    def exec_module(self, **kwargs):
+
+        for key in self.module_arg_spec:
```
```suggestion
        for key in list(self.module_arg_spec.keys()) + ['tags']:
```

Fred-sun commented on plugins/modules/azure_rm_privateendpointdnszonegroup_info.py:
```diff
@@ -0,0 +1,249 @@
+
+        super(AzureRMPrivateEndpointDnsZoneGroupInfo, self).__init__(self.module_arg_spec,
+                                                                     supports_check_mode=True,
+                                                                     supports_tags=False,
```
Since the document inherits tags, it is recommended that argc be set to true and the tags parameter can be passed. The resources of the specified tags can be filtered through tags. Thank you very much!
```suggestion
                                                                     supports_tags=True,
```

Fred-sun commented on plugins/modules/azure_rm_privateendpointdnszonegroup_info.py:
```diff
@@ -0,0 +1,249 @@
+
+EXAMPLES = '''
+- name: Get specific DNS zone groups for a private endpoing
+    azure_rm_privateendpointdnszonegroup_info:
```
```suggestion
  azure_rm_privateendpointdnszonegroup_info:
```

Fred-sun commented on plugins/modules/azure_rm_privateendpointdnszonegroup_info.py:
```diff
@@ -0,0 +1,249 @@
+    resource_group: "my-resource-group"
+
+- name: Get all DNS zone groups for a private endpoint
+    azure_rm_privateendpointdnszonegroup_info:
```
```suggestion
  azure_rm_privateendpointdnszonegroup_info:
```

Fred-sun commented on plugins/modules/azure_rm_privateendpointdnszonegroup_info.py:
```diff
@@ -0,0 +1,249 @@
+    description:
+        - List of private endpoint zone groups.
+    returned: always
+    type: list
```
```suggestion
    type: complex
```

Fred-sun commented on plugins/modules/azure_rm_privateendpointdnszonegroup_info.py:
```diff
@@ -0,0 +1,249 @@
+            description:
+                - List of zone configuration within the zone group.
+            returned: always
+            type: list
```
```suggestion
            type: complex
```

Fred-sun commented on plugins/modules/azure_rm_privateendpointdnszonegroup_info.py:
```diff
@@ -0,0 +1,249 @@
+                    description:
+                        - List of DNS records for zone.
+                    returned: always
+                    type: list
```
```suggestion
                    type: complex
```

Fred-sun commented on plugins/modules/azure_rm_privateendpointdnszonegroup_info.py:
```diff
@@ -0,0 +1,249 @@
+        self.log("List all in {0}".format(self.private_endpoint))
+        try:
+            items = self.network_client.private_dns_zone_groups.list(private_endpoint_name=self.private_endpoint, resource_group_name=self.resource_group)
+            
```
```suggestion

```

Fred-sun commented on plugins/modules/azure_rm_privateendpointdnszonegroup.py:
```diff
@@ -0,0 +1,351 @@
+            return self.zone_to_dict(item)
+        except Exception:
+            self.log("Did not find the private endpoint resource")
+        
```
```suggestion

```

Fred-sun commented on plugins/modules/azure_rm_privateendpointdnszonegroup.py:
```diff
@@ -0,0 +1,351 @@
+        id:
+            description:
+                - Resource ID of the private endpoint zone group.
+            sample: /subscriptions/xxx/resourceGroups/myResourceGroup/providers/Microsoft.Network/privateEndpoints/myPrivateEndpoint/privateDnsZoneGroups/myZoneGroup
```
line too long (165 > 160 characters)

Fred-sun commented on plugins/modules/azure_rm_privateendpointdnszonegroup.py:
```diff
@@ -0,0 +1,351 @@
+                        - ID of the private DNS zone.
+                    returned: always
+                    type: str
+                    sample: /subscriptions/xxx/resourceGroups/myResourceGroup/providers/Microsoft.Network/privateDnsZones/privatelink.postgres.database.azure.com
```
line too long (161 > 160 characters)

Fred-sun commented on plugins/modules/azure_rm_privateendpointdnszonegroup_info.py:
```diff
@@ -0,0 +1,249 @@
+        id:
+            description:
+                - Resource ID of the private endpoint zone group.
+            sample: /subscriptions/xxx/resourceGroups/myResourceGroup/providers/Microsoft.Network/privateEndpoints/myPrivateEndpoint/privateDnsZoneGroups/myZoneGroup
```
line too long (165 > 160 characters)

Fred-sun commented on plugins/modules/azure_rm_privateendpointdnszonegroup_info.py:
```diff
@@ -0,0 +1,249 @@
+                        - ID of the private DNS zone.
+                    returned: always
+                    type: str
+                    sample: /subscriptions/xxx/resourceGroups/myResourceGroup/providers/Microsoft.Network/privateDnsZones/privatelink.postgres.database.azure.com
```
line too long (161 > 160 characters)

l3ender committed:
Merge branch 'dev' into private-endpoint-dns-zones

l3ender committed:
Apply suggestions from code review

Co-authored-by: Fred-sun <37327967+Fred-sun@users.noreply.github.com>

l3ender committed:
Merge branch 'private-endpoint-dns-zones' of github.com:l3ender/azure into private-endpoint-dns-zones

l3ender committed:
remove documentation reference to tags

l3ender committed:
linting fixes

l3ender commented:
Thanks for the review, @Fred-sun. The reference to extending tag documentation was a mistake: this resource does not support tags. I have updated that and also fixed lint issues.

Please review again and let me know if there are any other items!

l3ender committed:
Merge branch 'dev' into private-endpoint-dns-zones

l3ender commented:
Hello, wondering if there is anything else I can do for this PR? Thank you!

Fred-sun commented:
@l3ender We are reviewing and will push forward the merger as soon as possible. Thank you very much!

xuzhang3 commented:
LGTM
